src/common/pico_sync/mutex.c

```diff
@@ -73,41 +129,56 @@ bool __time_critical_func(mutex_enter_timeout_us)(mutex_t *mtx, uint32_t timeout
 
 bool __time_critical_func(mutex_enter_block_until)(mutex_t *mtx, absolute_time_t until) {
     assert(mtx->core.spin_lock);
+    lock_owner_id_t caller = lock_get_caller_owner_id();
     do {
         uint32_t save = spin_lock_blocking(mtx->core.spin_lock);
-        lock_owner_id_t caller = lock_get_caller_owner_id();
-        if (mtx->owner == LOCK_INVALID_OWNER_ID) {
+        if (!lock_is_owner_id_valid(mtx->owner) || (mtx->owner == caller && mtx->recursive)) {
             mtx->owner = caller;
-        } else if (mtx->owner == caller && mtx->recursion_state > 1) {
-            mtx->recursion_state--;
+            uint __unused total = mtx->enter_count++;
+            spin_unlock(mtx->core.spin_lock, save);
+            assert(!total); // check for overflow
+            return true;
         } else {
             if (lock_internal_spin_unlock_with_best_effort_wait_or_timeout(&mtx->core, save, until)) {
                 // timed out
                 return false;
-            } else {
-                // not timed out; spin lock already unlocked, so loop again
-                continue;
             }
+            // not timed out; spin lock already unlocked, so loop again
         }
-        spin_unlock(mtx->core.spin_lock, save);
-        return true;
     } while (true);
 }
 
-void __time_critical_func(mutex_exit)(mutex_t *mtx) {
+static void __time_critical_func(mutex_exit_nr_guts)(mutex_t *mtx) {
     uint32_t save = spin_lock_blocking(mtx->core.spin_lock);
-    assert(mtx->owner != LOCK_INVALID_OWNER_ID);
-    if (!mtx->recursion_state) {
-        mtx->owner = LOCK_INVALID_OWNER_ID;
+    assert(lock_is_owner_id_valid(mtx->owner));
+    mtx->owner = LOCK_INVALID_OWNER_ID;
+    lock_internal_spin_unlock_with_notify(&mtx->core, save);
+}
+
+static void __time_critical_func(mutex_exit_r_guts)(mutex_t *mtx) {
+    uint32_t save = spin_lock_blocking(mtx->core.spin_lock);
+    assert(mtx->enter_count);
+    if (!--mtx->enter_count) {
         lock_internal_spin_unlock_with_notify(&mtx->core, save);
```

Comment:
Trying to compare the old-code with the new-code, is an `mtx->owner = LOCK_INVALID_OWNER_ID;` also needed before this line? Or is that not necessary now? :shrug:

Reply:
ha good catch, i was rather saddened to see there was no test currently (I did test it originally; i guess I'll add one)